engine/dlinked_list.hpp

```diff
@@ -4,373 +4,402 @@
 
 #pragma once
 
-#include <algorithm>
-#include <bitset>
 #include <cassert>
 #include <cstdint>
 #include <cstdio>
 #include <cstdlib>
+
+#include <algorithm>
+#include <bitset>
 #include <exception>
 #include <iomanip>
 #include <iostream>
 
 #include <libpmem.h>
 #include <libpmemobj++/string_view.hpp>
 
-#include "hash_table.hpp"
 #include "kvdk/engine.hpp"
+
+#include "alias.hpp"
+#include "hash_table.hpp"
+#include "macros.h"
 #include "structures.hpp"
 #include "utils.hpp"
 
-#include <list>
-
-#define hex_print(x)                                                           \
-  std::hex << std::setfill('0') << std::setw(sizeof(decltype(x)) * 2) << x
-
 namespace KVDK_NAMESPACE {
-/// DListIterator does not pin DlinkedList
-/// It's up to caller to ensure that any
-/// DListIterator constructed belongs to the
-/// correct DlinkedList and to ensure that
-/// DlinkedList is valid when DListIterator
-/// accesses data on it.
-class DListIterator {
-private:
-  friend class DLinkedList;
-  friend class UnorderedCollection;
-
-private:
-  /// PMem pointer to current Record
-  PMEMAllocator *p_pmem_allocator_;
-  /// Current position
-  DLRecord *pmp_curr_;
-
+/// DLinkedList is a helper class to access PMem
+/// DLinkedList guarantees that forward links are always valid
+/// Backward links may be broken
+/// if shutdown happens when new record is being emplaced.
+/// DLinkedList does not deallocate records.
+/// Deallocation is done by caller.
+template <RecordType HeadType, RecordType TailType, RecordType DataType>
+class DLinkedList {
 public:
-  /// It's up to caller to provide correct PMem pointer
-  /// and PMemAllocator to construct a DListIterator
-  explicit DListIterator(PMEMAllocator *p_pmem_allocator, DLRecord *curr)
-      : p_pmem_allocator_{p_pmem_allocator}, pmp_curr_{curr} {}
+  /// It's up to caller to ensure that any
+  /// iterator constructed belongs to the
+  /// correct DlinkedList and to ensure that
+  /// DlinkedList is valid when iterator
+  /// accesses data on it.
+  class iterator {
+  private:
+    friend class DLinkedList;
+
+  private:
```

Comment:
merge private？

Reply:
That's three different fields. 
One field holds friend class or functions.
One field holds internal variables.
One field holds a private constructor, which is strange, but user really shouldn't construct a iterator himself. Thus the constructor is private.